### ISSUE TITLE
no resizing method when deal consep datasets

### ISSUE THREAD

**Issue by Deng-Jiang-Lin** [Lewislou/SENUCLS#1]
你好，consep数据集无法在extract_patches.py提取出1024x1024的patch，论文中说明了需要将consep数据集resize到1024x1024 代码里面似乎没有这部分代码，劳驾作者解答以下，万分感谢。

**Comment by Lewislou** [comment]
HI,

You can resize the dataset and use the resized dataset in the patch extraction. For example, you may use the PIL library to resize the image, inst map and type map.

A example code shippet:
img = Image.open(image_path + name)
label = io.loadmat(label_path + name[:-4] + '.mat')
inst_map = label['inst_map']
type_map = label['type_map']
inst_map = Image.fromarray(inst_map) 
type_map = Image.fromarray(type_map) 
img = img.resize((1024, 1024),Image.ANTIALIAS)
inst_map = inst_map.resize((1024, 1024),Image.NEAREST)
type_map = type_map.resize((1024, 1024),Image.NEAREST)
img.save(save_image_path + name)
io.savemat(dataNew, {'inst_map':np.array(inst_map),'type_map':np.array(type_map)})

**Comment by Deng-Jiang-Lin** [comment]
非常感谢您的答复，按照您之前的描述我已经完成了resize工作，**再次感谢**！我目前**还有些困惑**。



1. 在测试consep数据集时，是按照config.py源**代码注释**将act_shape和out_shape调整为**[270,270]和 [80,80]**,还是按照论文**[1024，1024]**进行设置？（我在阅读源代码的时候，发现这个变量主要用于数据增强部分的裁剪。）我按照注释设置时候act_shape[270,270]和out_shape[80,80]时，代码出现报错，显示encoder最后两层维度不匹配。这似乎是在OverlapPatchEmbed中，每个encoder层根据img_size调整了尺寸，导致shape不匹配出现的问题。

   ```python
   for branch_name, branch_desc in self.decoder.items():
       #  将最后两层特征进行上采样，用于特征融合，可以利用不同尺度的特征信息,提高模型的表现能力
       u3 = self.upsample2x(d[-1]) + d[-2]  # TODO 为什么这里维度不一样
       u3 = branch_desc[0](u3)
   ```

   2. 在调整focal_loss的权重时，我通过分析数据计算出四种细胞类型的比例和细胞总数(24332)，consep数据集的各细胞类型比例为

      * other: 0.0383
      * inflammatory: 0.2293
      * healthy epithelial: 0.3596
      * dysplastic/malignant epithelial: 0.3728

      我将training_weights设置为**[1-ration]** 。但我看原代码设置权重是**整数**，这是否有影响？我是研一新生，很多东西还了解得不是很透彻。
      **麻烦作者在帮我解答一下，万分感谢！**

### English

Thank you very much for your reply, I have completed the resize work according to your previous description, **Thanks again**! I'm **still a little confused** at the moment.

1. When testing the consep data set, should we adjust act_shape and out_shape to **[270,270] and [80,80]** according to the config.py source **code comments**, or according to the paper **[1024, 1024]* *To set it up? (When I read the source code, I found that this variable is mainly used for cropping of the data enhancement part.) When I set act_shape[270,270] and out_shape[80,80] according to the comments, the code reported an error and displayed the last two dimensions of the encoder. Mismatch. This seems to be a problem in OverlapPatchEmbed where each encoder layer is resized according to img_size, causing shape mismatch.

```
	for branch_name, branch_desc in self.decoder.items():
    #  将最后两层特征进行上采样，用于特征融合，可以利用不同尺度的特征信息,提高模型的表现能力
    u3 = self.upsample2x(d[-1]) + d[-2]  # TODO 为什么这里维度不一样
    u3 = branch_desc[0](u3)
```

​			2. When adjusting the weight of focal_loss, I calculated the proportion of the four cell types and the total number 				of cells (24332) by analyzing the data. The proportion of each cell type in the consep data set is

* other: 0.0383
  *inflammatory: 0.2293
* healthy epithelial: 0.3596
* dysplastic/malignant epithelial: 0.3728

​			I set training_weights to **[1-ration]**. But I see that the original code sets the weight to an **integer**. Does this have 			any  impact? I am a first-year graduate student, and I still don’t understand many things thoroughly.

**Could you please help me answer this question, thank you very much!**

**Comment by Deng-Jiang-Lin** [comment]
您好，我在复现论文的时候，**模型始终无法收敛**，精度一直很低且不稳定，losss下降但很缓慢，最终我推理模型，发现效果与原论文有一定差距。根据文档我主要对代码进行了以下更改

1.将consep数据集重新resize到1024x1024后，通过extract_patches.py获得.npy文件，其中

```python
 win_size = [1024, 1024] # 窗口
 step_size = [1024, 1024] # 步长
```

因此.npy文件的中图像尺寸为1024x1024。resize代码为

```python
def resize_consep():
    """Resize CoNSeP dataset to 1024x1024."""
    root = r"E:\数据集\consep\CoNSeP"
    mode = "Test"  # 更改类型
    img_root = f"{root}/{mode}/Images"
    labels_root = f"{root}/{mode}/Labels"
    res_img_path = f"E:/数据集/consep_1024x1024/{mode}/Images"
    res_labels_path = f"E:/数据集/consep_1024x1024/{mode}/Labels"

    ##############################
    def rm_n_mkdir(dir_path):
        """Remove and make directory."""
        if os.path.isdir(dir_path):
            shutil.rmtree(dir_path)
        os.makedirs(dir_path)

    [rm_n_mkdir(root) for root in [res_img_path, res_labels_path]]
    # rm_n_mkdir(res_img_path)
    # rm_n_mkdir(res_labels_path)
    # resize图像
    imgs = [os.path.join(img_root, file) for file in os.listdir(img_root)]  # 获取路径
    for img in imgs:
        _img = Image.open(img)
        _img = _img.resize((1024, 1024), Image.LANCZOS)
        _img.save(os.path.join(res_img_path, os.path.basename(img)))
    # resize inst_map和type_map
    labels = [os.path.join(labels_root, file) for file in os.listdir(labels_root)]
    for label in labels:
        data = io.loadmat(label)
        inst_map = data['inst_map']
        type_map = data['type_map']
        inst_map = Image.fromarray(inst_map)
        type_map = Image.fromarray(type_map)
        inst_map = inst_map.resize((1024, 1024), Image.NEAREST)
        type_map = type_map.resize((1024, 1024), Image.NEAREST)

        io.savemat(os.path.join(res_labels_path, os.path.basename(label)),
                   {'inst_map': np.array(inst_map), 'type_map': np.array(type_map)})

    # 查看数据是否resize成功
    print("Checking data...")
    res_imgs = [os.path.join(res_img_path, file) for file in os.listdir(res_img_path)]
    res_labels = [os.path.join(res_labels_path, file) for file in os.listdir(res_labels_path)]
    assert len(res_imgs) == len(res_labels)
    for res_img, res_label in zip(res_imgs, res_labels):
        img = Image.open(res_img)
        label = io.loadmat(res_label)
        inst_map = label['inst_map']
        type_map = label['type_map']

        assert img.size == (1024, 1024)
        assert inst_map.shape == (1024, 1024)
        assert type_map.shape == (1024, 1024)
    print("Done!")
```

2.在consep当中，我主要做以下调整

```python
    act_shape = [1024, 1024]  # patch shape used as input to network - central crop 
    out_shape = [1024, 1024]  # patch shape at output of network
    nr_type = 5  # 4核+1背景
```

3.在run_desc.py中，我对权重进行了以下调整

```
training_weights = [60,20, 8, 15]
```

其中我计算出4类细胞的占比为:

- other: 0.0383
- inflammatory: 0.2293
- healthy epithelial: 0.3596
- dysplastic/malignant epithelial: 0.3728



## English

Hello, when I reproduced the paper, the **model never converged**, the accuracy was always low and unstable, and the loss dropped but slowly. Finally, I reasoned about the model and found that the effect was somewhat different from the original paper. According to the documentation I mainly made the following changes to the code

1. After resizing the consep data set to 1024x1024, obtain the .npy file through extract_patches.py, where

```python
 win_size = [1024, 1024] # 窗口
 step_size = [1024, 1024] # 步长
```

​              So the medium image size of the .npy file is 1024x1024. The resize code is

```python
	def resize_consep():
"""Resize CoNSeP dataset to 1024x1024."""
root = r"E:\数据集\consep\CoNSeP"
mode = "Test"  # 更改类型
img_root = f"{root}/{mode}/Images"
labels_root = f"{root}/{mode}/Labels"
res_img_path = f"E:/数据集/consep_1024x1024/{mode}/Images"
res_labels_path = f"E:/数据集/consep_1024x1024/{mode}/Labels"

##############################
def rm_n_mkdir(dir_path):
    """Remove and make directory."""
    if os.path.isdir(dir_path):
        shutil.rmtree(dir_path)
    os.makedirs(dir_path)

[rm_n_mkdir(root) for root in [res_img_path, res_labels_path]]
# rm_n_mkdir(res_img_path)
# rm_n_mkdir(res_labels_path)
# resize图像
imgs = [os.path.join(img_root, file) for file in os.listdir(img_root)]  # 获取路径
for img in imgs:
    _img = Image.open(img)
    _img = _img.resize((1024, 1024), Image.LANCZOS)
    _img.save(os.path.join(res_img_path, os.path.basename(img)))
# resize inst_map和type_map
labels = [os.path.join(labels_root, file) for file in os.listdir(labels_root)]
for label in labels:
    data = io.loadmat(label)
    inst_map = data['inst_map']
    type_map = data['type_map']
    inst_map = Image.fromarray(inst_map)
    type_map = Image.fromarray(type_map)
    inst_map = inst_map.resize((1024, 1024), Image.NEAREST)
    type_map = type_map.resize((1024, 1024), Image.NEAREST)

    io.savemat(os.path.join(res_labels_path, os.path.basename(label)),
               {'inst_map': np.array(inst_map), 'type_map': np.array(type_map)})

# 查看数据是否resize成功
print("Checking data...")
res_imgs = [os.path.join(res_img_path, file) for file in os.listdir(res_img_path)]
res_labels = [os.path.join(res_labels_path, file) for file in os.listdir(res_labels_path)]
assert len(res_imgs) == len(res_labels)
for res_img, res_label in zip(res_imgs, res_labels):
    img = Image.open(res_img)
    label = io.loadmat(res_label)
    inst_map = label['inst_map']
    type_map = label['type_map']

    assert img.size == (1024, 1024)
    assert inst_map.shape == (1024, 1024)
    assert type_map.shape == (1024, 1024)
print("Done!")
```

2. In consep, I mainly make the following adjustments

   ```python
   act_shape = [1024, 1024]  # patch shape used as input to network - central crop 
   out_shape = [1024, 1024]  # patch shape at output of network
   nr_type = 5  # 4核+1背景
   ```

3. 在run_desc.py中，我对权重进行了以下调整

```
	training_weights = [60,20, 8, 15]
```

Among them, I calculated the proportion of 4 types of cells as:

- other: 0.0383
  -inflammatory: 0.2293
- healthy epithelial: 0.3596
- dysplastic/malignant epithelial: 0.3728

**Comment by Lewislou** [comment]
Hi, 

What's your classification results? Could your provide some outputed logs in the training of stage 00? 
What's the nuclei segmentation method that you used and how is the detection F-score? 
Based on your calculated nuclei proportion, the weights for classification loss are [1/0.0383,1/0.2293,1/0.3596,1/0.3728] which can be converted into interger wights that close to [26,4,2,2]. These weights work well in my experiments.  

Hope this could help you~

**Comment by Deng-Jiang-Lin** [comment]
好的，我对consep数据集重新进行了一次实验。根据github训练步骤，我只进行了如下几部分更改

1.	config.py

```python
# shape into net
act_shape = [256, 256] 
out_shape = [256, 256] 
nr_type =5 # 4 types and 1 background
```

2. extract_patches.py

```python
# extract patch size
win_size = [1024, 1024]
step_size = [1024, 1024
```

3. dataset.py

```python
# merge classes for CoNSeP (in paper we only utilise 3 nuclei classes and background)
# If own dataset is used, then the below may need to be modified
# 在consep数据集中，我将原始细胞从7类归并为4类
ann_type[(ann_type == 3) | (ann_type == 4)] = 3
ann_type[(ann_type == 5) | (ann_type == 6) | (ann_type == 7)] = 4 
```

4. models.senucls.opt.py

```python
# 在训练分割部分使用预训练模型
"pretrained": "D:/paperCode/SENUCLS/models/pretrained/van_base_convert.pth.tar" 
# 两个分支使用的batch大小
"batch_size": {"train": 16, "valid": 16},  
```

5. run_desc.py

```python
training_weights = [26,4,2,2]
```

6. compute_stats.py

```
# in 'run_nuclei_type_stat_panuke' method
# * for converting the GT type in CoNSeP
true_inst_type[(true_inst_type == 3) | (true_inst_type == 4)] = 3
true_inst_type[(true_inst_type == 5) | (true_inst_type == 6) | (true_inst_type == 7)] = 4
```

7. type_info.json

```json
{   "0": ["background", [0,0,0]],
    "1" : ["other", [255,0,0]],
    "2" : ["inflammatory", [0,255,0]],
    "3" : ["epithelial", [0, 0, 255]],
    "4" : ["spindle-shaped", [255, 255, 0]]
}
```

实验结果如下：

1. 执行compute_stats.py

```python
[ 0.99977  0.08184  0.06514  0.00000  0.04536  0.00386]
f1 average: 0.02858937020149762
```

2. 训练过程损失

   ```python
   # epoch 28
   accu_al1 [0.10416666666666667]
   accu_per [0.9710,0.0,0.0425,0.0]
   # epoch 38
   accu_al1 [0.149685534]
   accu_per [1.0,0.0,0.08,0.006]
   # epoch 88
   accu_al1 [0.14409]
   accu_per [0.9861,0.0055,01211,0.0050]
   ```

3. 00阶段输出日志,我将输入的间隔调整为5次一输出

```json
{
  "5": {
    "train-loss_tp_bce": 3.2773356267526745,
    "train-loss_tp_dice": 4.323199779081346,
    "train-loss_tp_focal": 8.10791342509985,
    "train-overall_loss": 20.031648857712746,
    "train-lr-net": 0.0001
  },
  "10": {
    "train-loss_tp_bce": 3.036726451765291,
    "train-loss_tp_dice": 4.264751426725515,
    "train-loss_tp_focal": 7.627226322006669,
    "train-overall_loss": 19.193455851816694,
    "train-lr-net": 0.0001
  }
}
```

4. 01阶段输出,由于json字符串太长，我删除了一些

```json
{

  "10": {
    "train-loss_tp_bce": 2.82381686913231,
    "train-loss_tp_dice": 4.225800491471434,
    "train-loss_tp_focal": 7.047607174630986,
    "train-overall_loss": 16.650688877218652,
    "train-lr-net": 0.0001
  },
  "15": {
    "train-loss_tp_bce": 2.5992428707457416,
    "train-loss_tp_dice": 4.179355440378496,
    "train-loss_tp_focal": 6.85064821216273,
    "train-overall_loss": 15.60506855683699,
    "train-lr-net": 0.0001
  },

  "25": {
    "train-loss_tp_bce": 2.2022372753776125,
    "train-loss_tp_dice": 4.096368035901207,
    "train-loss_tp_focal": 6.335304241758649,
    "train-overall_loss": 13.81690722709344,
    "train-lr-net": 0.0001
  },

  "55": {
    "train-loss_tp_bce": 1.639679807995923,
    "train-loss_tp_dice": 3.9519309879269096,
    "train-loss_tp_focal": 5.820565892439735,
    "train-overall_loss": 11.666093881883482,
    "train-lr-net": 1.0000000000000002e-06
  },

  "65": {
    "train-loss_tp_bce": 1.5731698557534113,
    "train-loss_tp_dice": 3.9311870947404133,
    "train-loss_tp_focal": 5.623958223045578,
    "train-overall_loss": 11.280344645977253,
    "train-lr-net": 1.0000000000000002e-06
  },
  "100": {
    "train-loss_tp_bce": 1.491195051359771,
    "train-loss_tp_dice": 3.894043708675973,
    "train-loss_tp_focal": 5.582640548283978,
    "train-overall_loss": 10.993128916051107,
    "train-lr-net": 1.0000000000000002e-07
  }
}
```
![Uploading 微信图片_20240509163319.png…]()

**Comment by Deng-Jiang-Lin** [comment]
非常感谢作者百忙之中能抽出时间来帮助我解决疑问，这给我提供了非常大的帮助，再次感谢！！！
1.实验结果，我尝试过将最终输入模型的shape更改为1024x1024，由于尺寸太大，造成了内存溢出，只能将batch调整为4或者更小，在论文当中说的是将consep调整为1024，指的是最终输入模型的尺寸吗（config.py）中的shape参数？
2.细胞核分割方法，细胞核分割方法采用的是"HoVer-Net: Simultaneous Segmentation and Classification of Nuclei in Multi-Tissue Histology Images"

**Comment by Lewislou** [comment]
Hi,

For CoNSep dataset, I use 1024*1024 for all the shape configuration in config.py. If you opt to utilize patches of size 256x256, you should extract patches of this size instead of 1024x1024 and adjust the class weights according to the training logs. For instance, if the training accuracy for class 1 is exceptionally high while it is very low for the other three classes, you could decrease the weight for class 1 and increase the weights for the other classes.

Please feel free to ask any additional questions. Thank you for your interest in our work!

**Comment by Deng-Jiang-Lin** [comment]
### 谢谢作者的回复，感谢您的提醒。
在extract patch时尺寸虽然为1024x1024，但是经过数据增强之后最终输入的模型的尺寸为256x256，这是仅仅是中心区域的截取，产生了图像其他**区域并未输入网络**的问题。
我这就去调整实验参数，并展开下一次实验。**感谢作者的耐心回答！**

**Comment by Deng-Jiang-Lin** [comment]
您好 ，经过更改 我目前的size 为256x256 ，loss权重为 [24,8,4,7] 现在的实验结果为 
[ 0.99977  0.71412  0.30910  0.44013  0.75759  0.52301]
f1 average: 0.5074595257273239
其他参数都是按照论文设置，经过多次实验，依旧不能达到论文中的效果。代码都断点调试了一遍，都快背下来了 ，请问您是否还保留consep数据集训练的参数。救救孩子要崩溃了 呜呜.....

**Comment by Lewislou** [comment]
Hi,

As previously mentioned, we train and test the model using images of size 1024x1024. The larger training size may help the GCN capture global context information and enhance its feature representation capabilities.

If you want to access the weights, please send an email from your institutional email address to "weilou@link.cuhk.edu.cn", and include your personal information such as name, occupation, affiliation, and the specific intended use. We will assess your intended use and background, and provide the pre-trained weights as soon as we can.

Wish you all the best~
